pkg/providers/provider.go

```diff
@@ -0,0 +1,35 @@
+package providers
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"strings"
+
+	"github.com/flant/k8s-image-availability-exporter/pkg/providers/amazon"
+	"github.com/flant/k8s-image-availability-exporter/pkg/providers/k8s"
+	"github.com/google/go-containerregistry/pkg/authn"
+)
+
+type Provider interface {
+	GetAuthKeychain(registryStr string) (authn.Keychain, error)
+}
+
+type ProviderRegistry map[string]Provider
+
+func NewProviderChain(pullSecretsGetter func(image string) []corev1.Secret) ProviderRegistry {
+	return map[string]Provider{
+		"amazon": amazon.NewProvider(),
+		"k8s":    k8s.NewProvider(pullSecretsGetter),
+	}
+}
+
+type ImagePullSecretsFunc func(image string) []corev1.Secret
+
+func (p ProviderRegistry) GetAuthKeychain(registryStr string) (authn.Keychain, error) {
+	switch {
+	case strings.Contains(registryStr, "amazonaws.com"):
```

Comment:
lets user `url.Parse` here and than check the URL with more accurate regex `^([0-9]{12})\.dkr\.ecr\.[a-z0-9-]+\.amazonaws\.com(\.cn)?$`

Reply:
Done
But i used regexp only